### PR TITLE
HIVE-25645: Query-based compaction doesn't work when partition column type is boolean

### PR DESCRIPTION
**Pull request by veghlaci05**
What changes were proposed in this pull request?

This PR fixes the issue, when compaction fails on tables with boolean partition column.

@zabetak , @kgyrtkirk This PR affects some classes which are really not my table, and I'm unsure if my changes are fine. Could you please also review it?
Why are the changes needed?

Compaction must work properly on tables with boolean partition column as well.
Does this PR introduce any user-facing change?

No
How was this patch tested?

Using automated tests.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/SerializationUtilities.java [hunk 1]
```diff
@@ -810,31 +811,44 @@ private static void serializeObjectByKryo(Kryo kryo, Object plan, OutputStream o
...
+    } finally {
+      releaseKryo(kryo);
+    }
   }
 
   public static String serializeExpression(ExprNodeGenericFuncDesc expr) {
-    return new String(Base64.encodeBase64(serializeExpressionToKryo(expr)),
-        StandardCharsets.UTF_8);
+    return new String(Base64.encodeBase64(serializeObjectToKryo(expr)),
+            StandardCharsets.UTF_8);
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/SerializationUtilities.java:846]: nit: keep 4 spaces 😄

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/ppr/PartitionPruner.java [hunk 2]
```diff
@@ -498,7 +498,7 @@ private static PrunedPartitionList getPartitionsFromServer(Table tab, final Stri
    * @return true iff the partition pruning expression contains non-partition columns.
    */
   static private boolean pruneBySequentialScan(Table tab, List<Partition> partitions,
-      ExprNodeGenericFuncDesc prunerExpr, HiveConf conf) throws HiveException, MetaException {
+                                               ExprNodeDesc prunerExpr, HiveConf conf) throws HiveException, MetaException {
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/ppr/PartitionPruner.java:501]: nit: line continuation is 4 spaces

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java [hunk 3]
```diff
@@ -347,8 +347,16 @@ private void buildWhereClauseForInsert(StringBuilder query) {
 
       query.append(" where ");
       for (int i = 0; i < keys.size(); ++i) {
-        query.append(i == 0 ? "`" : " and `").append(keys.get(i).getName()).append("`='")
-            .append(vals.get(i)).append("'");
+        FieldSchema keySchema = keys.get(i);
+        boolean isBooleanKey = keySchema.getType().equalsIgnoreCase("boolean");
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java:351]: ColumnType.BOOLEAN_TYPE_NAME

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java [hunk 4]
```diff
@@ -347,8 +347,16 @@ private void buildWhereClauseForInsert(StringBuilder query) {
...
+        query.append(i == 0 ? "`" : " and `").append(keySchema.getName()).append("`=");
+        if (!isBooleanKey) {
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java:353]: nit: Maybe easier to read:
```
if (isBooleanKey) {
  query.append("'").append(vals.get(i)).append("'");
} else {
  query.append(vals.get(i));
}
```

File: ql/src/test/org/apache/hadoop/hive/metastore/TestMetastoreExpr.java [hunk 5]
```diff
@@ -183,12 +183,12 @@ public void checkExpr(int numParts,
...
     List<Partition> parts = new ArrayList<Partition>();
     client.listPartitionsByExpr(dbName, tblName,
-        SerializationUtilities.serializeExpressionToKryo(expr), null, (short)-1, parts);
+        SerializationUtilities.serializeObjectWithTypeInformation(expr), null, (short)-1, parts);
     assertEquals("Partition check failed: " + expr.getExprString(), numParts, parts.size());
 
     // check with partition spec as well
     PartitionsByExprRequest req = new PartitionsByExprRequest(dbName, tblName,
-            ByteBuffer.wrap(SerializationUtilities.serializeExpressionToKryo(expr)));
+            ByteBuffer.wrap(SerializationUtilities.serializeObjectWithTypeInformation(expr)));
```
Comment (pvary) [inline ql/src/test/org/apache/hadoop/hive/metastore/TestMetastoreExpr.java:191]: nit: funny that the indentation is different than the above one.
If we touch the line fix the indentation please

**Review by klcopp — APPROVED** [review]

**Pull request merged**